pkg/sql/plan/function/operator/coalesce.go

```diff
@@ -0,0 +1,286 @@
+	}
+
+	CoalesceFloat32 = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[float32](vs, proc, types.Type{Oid: types.T_float32})
+	}
+
+	CoalesceFloat64 = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[float64](vs, proc, types.Type{Oid: types.T_float64})
+	}
+
+	CoalesceBool = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[bool](vs, proc, types.Type{Oid: types.T_bool})
+	}
+
+	CoalesceDate = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[types.Date](vs, proc, types.Type{Oid: types.T_date})
+	}
+
+	CoalesceDateTime = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[types.Datetime](vs, proc, types.Type{Oid: types.T_datetime})
+	}
+
+	CoalesceVarchar = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceString(vs, proc, types.Type{Oid: types.T_varchar})
+	}
+
+	CoalesceChar = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceString(vs, proc, types.Type{Oid: types.T_char})
+	}
+
+	CoalesceDecimal64 = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[types.Decimal64](vs, proc, types.Type{Oid: types.T_decimal64})
+	}
+
+	CoalesceDecimal128 = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[types.Decimal128](vs, proc, types.Type{Oid: types.T_decimal128})
+	}
+
+	CoalesceTimestamp = func(vs []*vector.Vector, proc *process.Process) (*vector.Vector, error) {
+		return coalesceGeneral[types.Timestamp](vs, proc, types.Type{Oid: types.T_timestamp})
+	}
+)
+
+// CoalesceTypeCheckFn is type check function for coalesce operator
+func CoalesceTypeCheckFn(inputTypes []types.T, _ []types.T, ret types.T) bool {
+	l := len(inputTypes)
+	for i := 0; i < l; i++ {
+		if inputTypes[i] != ret && inputTypes[i] != types.T_any {
+			return false
+		}
+	}
+	return true
+}
+
+// coalesceGeneral is a general evaluate function for coalesce operator
+// when return type is uint / int / float / bool / date / datetime
+func coalesceGeneral[T NormalType](vs []*vector.Vector, proc *process.Process, t types.Type) (*vector.Vector, error) {
+	vecLen := vector.Length(vs[0])
+	returnIsScalar := vs[0].IsScalar()
+
+	rs, err := proc.AllocVector(t, int64(vecLen*t.Oid.TypeLen()))
+	if err != nil {
+		return nil, err
+	}
+	rs.Col = vector.DecodeFixedCol[T](rs, t.Oid.TypeLen())
+	rs.Col = rs.Col.([]T)[:vecLen]
+	rsCols := rs.Col.([]T)
+
+	rs.Nsp = nulls.NewWithSize(vecLen)
+	rs.Nsp.Np.AddRange(0, uint64(vecLen))
+
+	for i := 0; i < len(vs); i++ {
+		input := vs[i]
+		cols := vector.MustTCols[T](input)
+		if input.IsScalar() {
+			if input.IsScalarNull() {
+				continue
+			}
+
+			if returnIsScalar {
```

Comment:
move it out of the loop and before you call the proc.AllocVector ?

Reply:
good point